File: agent-runtime/src/korean.ts
/**
 * Korean particle agreement, for sentences composed from a name the runtime did not choose.
 *
 * <b>This is not cosmetic, and the repository has now paid for it three times.</b> "쿠팡는"
 * (Cross-Channel v1), "쿠팡가 매출의 60%" (Demo Core Experience, backend) and, in this file's first
 * caller, the literal placeholder "은(는)" printed straight at a seller — each one a particle glued to
 * data without looking at the data. A sentence with the wrong particle reads as machine output, which
 * is precisely the impression an operations assistant cannot afford.
 *
 * <b>Shared, not per-module.</b> It lived inside `group/ChannelCoverage.ts` and inflected only channel
 * names, so the specialist writing product sentences reached for a placeholder instead. One rule, one
 * home, every caller.
 *
 * The rule: a Hangul syllable is `0xAC00 + 초성×588 + 중성×28 + 종성`, so `(code − 0xAC00) % 28 === 0`
 * means no final consonant. A name ending in anything else — a digit, a Latin letter — keeps the
 * open-syllable form, which is what a reader supplies for a foreign word anyway.
 */

function hasFinalConsonant(noun: string): boolean {
  const trimmed = noun.trim();
  if (trimmed.length === 0) {
    return false;
  }
  const code = trimmed.charCodeAt(trimmed.length - 1);
  if (code < 0xac00 || code > 0xd7a3) {
    return false;
  }
  return (code - 0xac00) % 28 !== 0;
}

/** "쿠팡" → "쿠팡은"; "네이버 스마트스토어" → "네이버 스마트스토어는". Topic particle. */
export function withTopic(noun: string): string {
  return noun.trim().length === 0 ? "" : `${noun}${hasFinalConsonant(noun) ? "은" : "는"}`;
}

/** "쿠팡" → "쿠팡이"; "스토어" → "스토어가". Subject particle. */
export function withSubject(noun: string): string {
  return noun.trim().length === 0 ? "" : `${noun}${hasFinalConsonant(noun) ? "이" : "가"}`;
}

File: backend/src/main/java/com/sellerops/common/Korean.java
package com.sellerops.common;

/**
 * Korean particle agreement, for sentences the product composes from a name it did not choose.
 *
 * <p><b>This exists because the product has now got it wrong twice in front of a live read.</b>
 * "쿠팡는" (Cross-Channel v1) and "쿠팡가 매출의 60%" (this package) were both a Korean particle
 * concatenated to a channel name without looking at the name. A seller reads that as software that
 * does not speak their language, and the fix is four lines of Hangul arithmetic rather than a lookup
 * table per channel that would be wrong again the day a channel is added.
 *
 * <p>The rule: a Hangul syllable is {@code 0xAC00 + (초성×588) + (중성×28) + 종성}, so
 * {@code (code − 0xAC00) % 28 == 0} means the syllable has no final consonant. A name ending in
 * anything else — a digit, a Latin letter, a bracket — is left to the caller's default rather than
 * guessed at, because the honest answer for "Cafe24" depends on how the seller says the number.
 */
public final class Korean {

    private static final char HANGUL_FIRST = 0xAC00;
    private static final char HANGUL_LAST = 0xD7A3;
    private static final int JONGSEONG_COUNT = 28;

    private Korean() {
    }

    /** True when the last character is a Hangul syllable carrying a final consonant (받침). */
    public static boolean endsWithFinalConsonant(String word) {
        if (word == null || word.isEmpty()) {
            return false;
        }
        char last = word.charAt(word.length() - 1);
        if (last < HANGUL_FIRST || last > HANGUL_LAST) {
            // Not a Hangul syllable. No claim is made; the caller's default particle stands.
            return false;
        }
        return (last - HANGUL_FIRST) % JONGSEONG_COUNT != 0;
    }

    /** {@code 쿠팡} → {@code 쿠팡은}; {@code 카페24 자사몰} → {@code 카페24 자사몰은}. Topic particle. */
    public static String withTopic(String word) {
        // A particle with nothing in front of it is not a shorter sentence, it is a stray syllable.
        return isBlank(word) ? "" : word + (endsWithFinalConsonant(word) ? "은" : "는");
    }

    /** {@code 쿠팡} → {@code 쿠팡이}; {@code 스토어} → {@code 스토어가}. Subject particle. */
    public static String withSubject(String word) {
        return isBlank(word) ? "" : word + (endsWithFinalConsonant(word) ? "이" : "가");
    }

    private static boolean isBlank(String word) {
        return word == null || word.isBlank();
    }
}
